lib/connection/connection.js

```diff
@@ -591,34 +594,87 @@ function makeUnsecureConnection(self, family) {
     connection_options = { port: self.port, host: self.host };
     connection_options.family = family;
   }
-
   const connection = net.createConnection(connection_options);
 
   // Set the options for the connection
   connection.setKeepAlive(self.keepAlive, self.keepAliveInitialDelay);
   connection.setTimeout(self.connectionTimeout);
   connection.setNoDelay(self.noDelay);
 
-  connection.once('connect', function() {
-    // Set socket timeout instead of connection timeout
-    connection.setTimeout(self.socketTimeout);
-    // Emit connect event
-    self.emit('connect', self);
-  });
-
   return connection;
 }
 
+function errorHandlerForIPv6(self) {
+  return function(err) {
+    if (connectionAccounting) deleteConnection(self.id);
+    // Debug information
+    if (self.logger.isDebug())
+      self.logger.debug(
+        f(
+          'connection %s for [%s:%s] errored out with [%s]',
+          self.id,
+          self.host,
+          self.port,
+          JSON.stringify(err)
+        )
+      );
+  };
+}
+
+function onConnect(self) {
+  self.connection.setTimeout(self.socketTimeout);
+  self.emit('connect', self);
+}
+
+function fastFallback(self, family, _options, _errorHandler) {
+  const _errorHandlerForIPv6 = errorHandlerForIPv6(self);
+  let connected = false;
+  if (family === 6 && !self.domainSocket && self.family === void 0) {
+    self.ipv6Connection = makeUnsecureConnection(self, 6);
+    setupConnection(self, self.ipv6Connection, _errorHandlerForIPv6);
+
+    self.ipv4Connection = makeUnsecureConnection(self, 4);
+    setupConnection(self, self.ipv4Connection, _errorHandler);
+
+    self.ipv6Connection.once('connect', function() {
+      if (!connected) {
+        self.connection = self.ipv6Connection;
+        onConnect(self);
```

Comment:
I originally ended the ipv4 connection and removed the listeners here, but this was causing a timeout test to fail because the connection timeout event would not be emitted. However, now I do not know where to shutdown the connection and remove the listeners.